projecttacoma/cql-exec-fhir
# Async Patient Source

# Summary
Added functionality to handle an asynchronous patient source which queries a FHIR server to retrieve patient data and records

## New Behavior
This PR sets up cql-exec-fhir to process async PatientSources which will eventually be able to get passed in after updates to fqm-execution. This PR does the legwork in cql-exec-fhir that will eventual enable users to pass in a simple list of patientIDs and a fhir server url though fqm-execution for measure evaluation, data requirements, and gaps in care, rather than an entire patient bundle or full fleshed out patient source.

## Code Changes
- Added `AsyncPatientSource` class which creates an axios instance with the passed in server url and defines the asynchronous functions `currentPatient()` and `nextPatient()` which are called by cql-execution during eCQM calculation
- Added `AsyncPatient` class which gets returned by `currentPatient()` and `nextPatient()`, stores FHIR patient data, and implements an asynchronous version of `findRecords()` which will pull all the resources of a given type that reference the patient
- Added Patient compartment definition to identify how resources can reference patient
- Updated babelrc to correctly compile using async/await terminology
- Added Prettier
- Updated eslint config to use ECMA 12
- Added unit testing

# Testing Guidance
Dang prettier really did a number on these files... those changes can largely be ignored. The code that actually needs a rigorous review is the AsyncPatientSource and AsyncPatient classes in fhir.js as well as the changes to package.json, the compartment definition file, and the prettier + eslint + babel config changes.

These changes can be integration tested by making a few changes to fqm execution:
- Add the following classes to `src/types/CQLExecFHIR.d.ts`:
```  
  class AsyncPatient extends FHIRObject implements IPatient {
    findRecords(profile: string): FHIRObject;
  }

  export class AsyncPatientSource {
    constructor(serverInfo: string);
    loadPatientIds(ids: string[]): void;
    currentPatient(): AsyncPatient | undefined;
    nextPatient(): AsyncPatient | undefined;
    reset: void;
    static FHIRv401(serverInfo: string): AsyncPatientSource;
  }
```
- Add the following at line 131 of `cli.ts`: 
```
  import * as cqlExecFhir from 'cql-exec-fhir';
  const TEST_IDS = ['numer-EXM130', 'denom-EXM130'];
  const patientSource = cqlExecFhir.AsyncPatientSource.FHIRv401('http://localhost:3000/4_0_1');
  patientSource.loadPatientIds(TEST_IDS);
```
- Add `await` before `executor.exec` on line 101 of `src/execution/Execution.ts`. The line should now read: 
`const results = await executor.exec(patientSource, executionDateTime);`
- Copy `cql-execution` from repository [here](https://github.com/projecttacoma/cql-execution)
- Run `npm link path/to/cql-execution/directory`
- Run `npm link path/to/cql-exec-fhir/directory`
- Run `npm run build`
- Clear your test server database by running `npm run db:reset`
- Start the test server at `http://localhost:3000`
- Upload the EXM130 transaction bundle to the test server. Slack me if you don't know how to do this
- Run `node build/cli.ts  -m "path/to/EXM130/Measure/bundle" -o "reports" -p "path/to/EXM130/patient/bundle" -s "2019-01-01" -e "2019-12-31" -f > output.json`
- Note, the -f flag here will cause fqm-execution to compile and AsyncPatientSource instead of using in the pased-in patient bundle, however an error will be thrown if the patientBundles arg is entirely omitted
- Check the ouput.json file and ensure that we have exactly 1 numerator and 1 denominator patient :)
- If you wanna be super rigorous, check the test server logs and ensure that it is searching for patients and other fhir resources during the measure calculation.
